### PR TITLE
layout: make `ContainingBlock` use `Au` for `inline_size` and `block_size`

### PR DESCRIPTION
**Pull request by atbrakhi**
Make ContainingBlock use Au for inline_size and block_size

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

### CODE REVIEW

File: components/layout_2020/flow/mod.rs [hunk 1]
```diff
@@ -141,18 +141,20 @@ impl BlockLevelBox {
...
-                containing_block.inline_size -
-                    pbm.padding_border_sums.inline.into() -
-                    margin_inline_start -
-                    margin_inline_end
+                Length::from(
+                    containing_block.inline_size -
+                        pbm.padding_border_sums.inline -
+                        margin_inline_start.into() -
+                        margin_inline_end.into(),
+                )
```
Comment (Loirooriol) [inline components/layout_2020/flow/mod.rs:149]: Can avoid the `into()`:
```suggestion
                Length::from(containing_block.inline_size - pbm.padding_border_sums.inline) -
                    margin_inline_start -
                    margin_inline_end
```

File: components/layout_2020/flow/mod.rs [hunk 2]
```diff
@@ -1241,10 +1241,12 @@ fn solve_containing_block_padding_border_and_margin_for_in_flow_box<'a>(
...
-            containing_block.inline_size -
-                pbm.padding_border_sums.inline.into() -
-                margin_inline_start -
-                margin_inline_end
+            Length::from(
+                containing_block.inline_size -
+                    pbm.padding_border_sums.inline -
+                    margin_inline_start.into() -
+                    margin_inline_end.into(),
+            )
```
Comment (Loirooriol) [inline components/layout_2020/flow/mod.rs:1249]: Ditto:

```suggestion
            Length::from(containing_block.inline_size - pbm.padding_border_sums.inline) -
                margin_inline_start -
                margin_inline_end
```

File: components/layout_2020/flow/mod.rs [hunk 3]
```diff
@@ -1337,8 +1339,9 @@ fn solve_inline_margins_avoiding_floats(
         (LengthOrAuto::Auto, _) => inline_adjustment + free_space,
         _ => inline_adjustment,
     };
-    let margin_inline_end = containing_block.inline_size - inline_size - margin_inline_start;
-    (margin_inline_start, margin_inline_end)
+    let margin_inline_end =
+        containing_block.inline_size - inline_size.into() - margin_inline_start.into();
+    (margin_inline_start, margin_inline_end.into())
```
Comment (Loirooriol) [inline components/layout_2020/flow/mod.rs:1344]:
```suggestion
    let margin_inline_end =
        Length::from(containing_block.inline_size) - inline_size - margin_inline_start;
    (margin_inline_start, margin_inline_end)
```

File: components/layout_2020/flow/mod.rs [hunk 4]
```diff
@@ -1353,15 +1356,18 @@ fn solve_inline_margins_for_in_flow_block_level(
...
-            Length::zero().max(free_space - end)
+            Length::zero().max(Length::from(free_space - end.into()))
         },
         (LengthOrAuto::LengthPercentage(start), _) => start,
     };
-    (margin_inline_start, free_space - margin_inline_start)
+    (
+        margin_inline_start,
+        Length::from(free_space - margin_inline_start.into()),
+    )
```
Comment (Loirooriol) [inline components/layout_2020/flow/mod.rs:1370]: Better keep `free_space` as a Length since it will need to be used as a Length:

```rs
    let free_space =
        Length::from(containing_block.inline_size - pbm.padding_border_sums.inline) - inline_size;
    let margin_inline_start = match (pbm.margin.inline_start, pbm.margin.inline_end) {
        (LengthOrAuto::Auto, LengthOrAuto::Auto) => Length::zero().max(free_space / 2.),
        (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(end)) => {
            Length::zero().max(free_space - end)
        },
        (LengthOrAuto::LengthPercentage(start), _) => start,
    };
    (margin_inline_start, free_space - margin_inline_start)
```

File: components/layout_2020/lib.rs [hunk 5]
```diff
@@ -33,7 +34,7 @@ use style::values::computed::{Length, LengthOrAuto};
 use crate::geom::LogicalVec2;
 
 pub struct ContainingBlock<'a> {
-    inline_size: Length,
+    inline_size: Au,
     block_size: LengthOrAuto,
```
Comment (Loirooriol) [inline components/layout_2020/lib.rs:38]: Seems a bit strange to have an Au in one axis, but a LengthOrAuto in the other.
Would it add much complexity to simultaneously change block_size into AuOrAuto?

Comment (atbrakhi) [reply components/layout_2020/lib.rs:38]: Earlier, I wasn't sure about how we wanted to proceed with the usage of `LengthOrAuto`, `LengthPercentage`, and `LengthPercentageOrAuto` with `Au`. That's why I didn't push those changes in this PR, but I do have a [wip branch](https://github.com/atbrakhi/servo/tree/make_block_size_use_au) that is making `block_size` use `AuOrAuto`. It still needs some work.

If you prefer, I can push those changes in this PR and we can do a wpt run, or I can open a separate PR for that?

File: components/layout_2020/flow/text_run.rs [hunk 6]
```diff
@@ -283,7 +283,7 @@ impl TextRun {
         let mut current: Option<(TextRunSegment, FontRef)> = None;
         let mut results = Vec::new();
 
-        let text = std::mem::replace(&mut self.text, String::new());
+        let text = std::mem::take(&mut self.text);
```
Comment (Loirooriol) [inline components/layout_2020/flow/text_run.rs:286]: This is an unrelated suggestion from clippy. It would be fine if you had to do other changes in this file, but otherwise I think it's better not to touch the file in this commit.

File: components/layout_2020/flow/text_run.rs [hunk 7]
```diff
@@ -331,7 +331,7 @@ impl TextRun {
                 // segment in the middle of the run (ie the start should be 0).
                 let start_byte_index = match current {
                     Some(_) => ByteIndex(current_byte_index as isize),
-                    None => ByteIndex(0 as isize),
+                    None => ByteIndex(0_isize),
```
Comment (Loirooriol) [inline components/layout_2020/flow/text_run.rs:334]: Ditto

File: components/layout_2020/lib.rs (hunk 5 above)
Comment (Loirooriol) [reply components/layout_2020/lib.rs:38]: I think it's a bit better to do them together, but not a big deal if you prefer to do them separately.

**Review by Loirooriol — APPROVED** [review]

File: components/layout_2020/flow/text_run.rs (hunk 6 above)
Comment (atbrakhi) [reply components/layout_2020/flow/text_run.rs]: Sorry about this, I thought i had removed these cahnges

File: components/layout_2020/lib.rs (hunk 5 above)
Comment (atbrakhi) [reply components/layout_2020/lib.rs]: I had finished those changes already so pushed it to this PR. Let see how wpt feels about these changes!

**Review by Loirooriol — APPROVED** [review]

**Comment by Loirooriol** [comment]
You should update the PR title since now it's not just `inline_size`. And squash the commits together with that title.
